### PR TITLE
Introduce Crypto Policy

### PR DESCRIPTION
**Pull request by tiraboschi**
Introduce cluster-wide Crypto Policy

Implement cluster-wide Crypto Policy API
as described in
https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/docs/enhancements/crypto-api.md

Propagate it to CNAO which is its first
consumer. Other Kubevirt Hyperconverged components
are going to follow in future PRs.

Signed-off-by: Simone Tiraboschi <stirabos@redhat.com>

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly
  
**Release note**:

```release-note
Introduce cluster-wide Crypto Policy
```

### CODE REVIEW

File: controllers/hyperconverged/hyperconverged_controller_test.go [hunk 1]
```diff
@@ -845,6 +848,132 @@ var _ = Describe("HyperconvergedController", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+				Expect(res).Should(Equal(reconcile.Result{}))
+
+				foundResource := &hcov1beta1.HyperConverged{}
+				Expect(
+					cl.Get(context.TODO(),
+						types.NamespacedName{Name: expected.hco.Name, Namespace: expected.hco.Namespace},
+						foundResource),
+				).To(BeNil())
```
Comment (nunnatsa) [inline controllers/hyperconverged/hyperconverged_controller_test.go:943]: To not add new cases after #1842 - please change to `ToNot(HaveOccurred())`.

File: controllers/hyperconverged/hyperconverged_controller_test.go [hunk 2]
```diff
@@ -845,6 +848,132 @@ var _ = Describe("HyperconvergedController", func() {
...
+				res, err = r.Reconcile(context.TODO(), rq)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(res.Requeue).To(BeFalse())
+				Expect(res).Should(Equal(reconcile.Result{}))
+
+				Expect(
+					cl.Get(context.TODO(),
+						types.NamespacedName{Name: expected.hco.Name, Namespace: expected.hco.Namespace},
+						foundResource),
+				).To(BeNil())
```
Comment (nunnatsa) [inline controllers/hyperconverged/hyperconverged_controller_test.go:969]: same

File: controllers/operands/networkAddons_test.go [hunk 3]
```diff
@@ -789,6 +791,77 @@ var _ = Describe("CNA Operand", func() {
...
+
+				// now, modify HCO's TLSSecurityProfile
+				hco.Spec.TLSSecurityProfile = modernTLSSecurityProfile
+
+				cl := commonTestUtils.InitClient([]runtime.Object{hco, existingResource})
+				handler := (*genericOperand)(newCnaHandler(cl, commonTestUtils.GetScheme()))
+				res := handler.ensure(req)
+				Expect(res.UpgradeDone).To(BeFalse())
+				Expect(res.Updated).To(BeTrue())
+				Expect(res.Err).To(BeNil())
```
Comment (nunnatsa) [inline controllers/operands/networkAddons_test.go:818]: `ToNot(HaveOccurred())`

File: controllers/operands/networkAddons_test.go [hunk 4]
```diff
@@ -789,6 +791,77 @@ var _ = Describe("CNA Operand", func() {
...
+
+				foundResource := &networkaddonsv1.NetworkAddonsConfig{}
+				Expect(
+					cl.Get(context.TODO(),
+						types.NamespacedName{Name: existingResource.Name, Namespace: existingResource.Namespace},
+						foundResource),
+				).To(BeNil())
```
Comment (nunnatsa) [inline controllers/operands/networkAddons_test.go:825]: `isTriggeredByApiServerCR`

File: controllers/operands/networkAddons_test.go [hunk 5]
```diff
@@ -789,6 +791,77 @@ var _ = Describe("CNA Operand", func() {
...
+				// now, modify CNAO node placement
+				existingResource.Spec.TLSSecurityProfile = modernTLSSecurityProfile
+
+				cl := commonTestUtils.InitClient([]runtime.Object{hco, existingResource})
+				handler := (*genericOperand)(newCnaHandler(cl, commonTestUtils.GetScheme()))
+				res := handler.ensure(req)
+				Expect(res.UpgradeDone).To(BeFalse())
+				Expect(res.Updated).To(BeTrue())
+				Expect(res.Overwritten).To(BeTrue())
+				Expect(res.Err).To(BeNil())
```
Comment (nunnatsa) [inline controllers/operands/networkAddons_test.go:849]: `isTriggeredByApiServerCR`

File: controllers/operands/networkAddons_test.go [hunk 6]
```diff
@@ -789,6 +791,77 @@ var _ = Describe("CNA Operand", func() {
...
+
+				foundResource := &networkaddonsv1.NetworkAddonsConfig{}
+				Expect(
+					cl.Get(context.TODO(),
+						types.NamespacedName{Name: existingResource.Name, Namespace: existingResource.Namespace},
+						foundResource),
+				).To(BeNil())
```
Comment (nunnatsa) [inline controllers/operands/networkAddons_test.go:856]: `isTriggeredByApiServerCR`

File: controllers/hyperconverged/hyperconverged_controller_test.go (hunk 1 above)
Comment (tiraboschi) [reply controllers/hyperconverged/hyperconverged_controller_test.go:943]: done

File: controllers/hyperconverged/hyperconverged_controller_test.go (hunk 2 above)
Comment (tiraboschi) [reply controllers/hyperconverged/hyperconverged_controller_test.go:969]: done

File: controllers/operands/networkAddons_test.go (hunk 3 above)
Comment (tiraboschi) [reply controllers/operands/networkAddons_test.go:818]: done

File: controllers/operands/networkAddons_test.go (hunk 4 above)
Comment (tiraboschi) [reply controllers/operands/networkAddons_test.go:825]: done

File: controllers/operands/networkAddons_test.go (hunk 5 above)
Comment (tiraboschi) [reply controllers/operands/networkAddons_test.go:849]: done

File: controllers/operands/networkAddons_test.go (hunk 6 above)
Comment (tiraboschi) [reply controllers/operands/networkAddons_test.go:856]: done

**Comment by tiraboschi** [comment]
ignoring slightly decreased coverage
/override coverage/coveralls

File: controllers/hyperconverged/hyperconverged_controller.go [hunk 7]
```diff
@@ -203,6 +206,30 @@ func add(mgr manager.Manager, r reconcile.Reconciler, ci hcoutil.ClusterInfo) er
...
+				log.Info(msg)
+				return []reconcile.Request{
+					{NamespacedName: apiServerCRPlaceholder},
+				}
+			}),
+		)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (nunnatsa) [inline controllers/hyperconverged/hyperconverged_controller.go:232]: Please consider to take it out to a new function, as sonarCloud warn about a code smell here (too complex function).

File: controllers/hyperconverged/hyperconverged_controller.go [hunk 8]
```diff
@@ -244,6 +271,19 @@ func (r *ReconcileHyperConverged) Reconcile(ctx context.Context, request reconci
...
+			return reconcile.Result{}, err
+		}
+		if apiServerCRTriggered {
+			logger.Info("Triggered by ApiServer CR, refreshing it")
+			err = hcoutil.GetClusterInfo().RefreshAPIServerCR(hcoRequest.Ctx, r.client)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+
```
Comment (nunnatsa) [inline controllers/hyperconverged/hyperconverged_controller.go:286]: SonarCloud warns about a code smell here (too complex function). Please consider to refactor.

Comment (tiraboschi) [reply controllers/hyperconverged/hyperconverged_controller.go]: done

**Comment by tiraboschi** [comment]
ignoring slightly decreased coverage
/override coverage/coveralls

**Pull request merged**